app/models/foreman_tasks/task/dynflow_task.rb

```diff
@@ -99,6 +99,30 @@ def running_steps
       execution_plan.try(:steps_in_state, :running, :suspended) || []
     end
 
+    def input_output_failed_steps
+      failed_steps.map do |f|
+        {
+          error: ({ exception_class: f.error.exception_class, message: f.error.message, backtrace: f.error.backtrace } if f.error),
+          action_class: f.action_class.name,
+          state: f.state,
+          input: f.action(execution_plan).input.pretty_inspect,
+          output: f.action(execution_plan).output.pretty_inspect
+        }.compact
+      end
+    end
+
+    def input_output_running_steps
+      running_steps.map do |f|
+        {
+          action_class: f.action_class.name,
+          state: f.state,
+          input: f.action(execution_plan).input.pretty_inspect,
+          output: f.action(execution_plan).output.pretty_inspect,
+          cancellable: cancellable_action?(f.action(execution_plan))
```

Comment:
`f.action(execution_plan)` does a query in the db, could you rewrite this in a way that the query is done only once per step?

Reply:
added a variable for it